ejb3/src/main/java/org/jboss/as/ejb3/remote/EJBRemotingConnectorClientMappingsEntryProviderService.java

```diff
@@ -91,24 +93,36 @@ public ServiceConfigurator configure(OperationContext context) {
         return new AbstractMap.SimpleImmutableEntry<>(this.group.get().getLocalMember().getName(), this.getClientMappings());
     }
 
+    /**
+     * This method provides client-mapping entries for all connected EJB clients.
+     * It returns either a set of user-defined client mappings for a multi-homed host or a single default client mapping for the single-homed host.
+     * Hostnames are preferred over literal IP addresses for the destination address part (due to EJBCLIENT-349).
+     *
+     * @return the client mappings for this host
+     */
     List<ClientMapping> getClientMappings() {
         final List<ClientMapping> ret = new ArrayList<>();
         RemotingConnectorBindingInfoService.RemotingConnectorInfo info = this.remotingConnectorInfo.get();
+
         if (info.getSocketBinding().getClientMappings() != null && !info.getSocketBinding().getClientMappings().isEmpty()) {
             ret.addAll(info.getSocketBinding().getClientMappings());
         } else {
-            // TODO: We use the textual form of IP address as the destination address for now.
-            // This needs to be configurable (i.e. send either host name or the IP address). But
-            // since this is a corner case (i.e. absence of any client-mappings for a socket binding),
-            // this should be OK for now
-            final String destinationAddress = info.getSocketBinding().getAddress().getHostAddress();
+            // for the destination, prefer the hostname over the literal IP address
+            final InetAddress destination = info.getSocketBinding().getAddress();
+            final String destinationName = Inet.toURLString(destination, true);
```

Comment:
This will wrap an IPv6 address literal with [ ] which is different from the previous code. Will the consumer handle that correctly?

Reply:
@rachmatowicz ^^^

Reply:
I'm pretty sure that I did check that aspect, Brian. The host part gets unpacked in EJBClientChannel in the EJB client library. IIRC, I did run the testsuite with IPv6 addresses and the results looked correct.